src/main/java/baritone/process/elytra/NetherPathfinderContext.java

```diff
@@ -24,98 +24,153 @@
+        else dim = NetherPathfinder.DIMENSION_END;
+        int height = Math.min(world.dimensionType().height(), 384);
+        if (!Baritone.settings().elytraAllowAboveRoof.value && dim == NetherPathfinder.DIMENSION_NETHER) height = Math.min(height, 128);
+        this.context = NetherPathfinder.newContext(seed, cache.toString(), dim, height, Baritone.settings().elytraCustomAllocator.value);
         this.seed = seed;
-        this.executor = Executors.newSingleThreadExecutor();
     }
 
     public boolean hasChunk(ChunkPos pos) {
         return NetherPathfinder.hasChunkFromJava(this.context, pos.x, pos.z);
     }
 
     public void queueCacheCulling(int chunkX, int chunkZ, int maxDistanceBlocks, BlockStateOctreeInterface boi) {
-        this.executor.execute(() -> {
-            synchronized (this.cullingLock) {
+        this.writeExecutor.execute(() -> {
+            writeLock.lock();
+            try {
                 boi.chunkPtr = 0L;
                 NetherPathfinder.cullFarChunks(this.context, chunkX, chunkZ, maxDistanceBlocks);
+            } finally {
+                writeLock.unlock();
             }
         });
     }
 
     public void queueForPacking(final LevelChunk chunkIn) {
         final SoftReference<LevelChunk> ref = new SoftReference<>(chunkIn);
-        this.executor.execute(() -> {
+        this.writeExecutor.execute(() -> {
             // TODO: Prioritize packing recent chunks and/or ones that the path goes through,
             //       and prune the oldest chunks per chunkPackerQueueMaxSize
             final LevelChunk chunk = ref.get();
             if (chunk != null) {
-                long ptr = NetherPathfinder.getOrCreateChunk(this.context, chunk.getPos().x, chunk.getPos().z);
-                writeChunkData(chunk, ptr);
+                writeLock.lock();
+                try {
+                    long ptr = NetherPathfinder.allocateAndInsertChunk(this.context, chunk.getPos().x, chunk.getPos().z);
+                    writeChunkData(chunk, ptr);
+                } finally {
+                    writeLock.unlock();
+                }
             }
         });
     }
 
     public void queueBlockUpdate(BlockChangeEvent event) {
-        this.executor.execute(() -> {
+        this.readExecutor.execute(() -> {
             ChunkPos chunkPos = event.getChunkPos();
-            long ptr = NetherPathfinder.getChunkPointer(this.context, chunkPos.x, chunkPos.z);
-            if (ptr == 0) return; // this shouldn't ever happen
-            event.getBlocks().forEach(pair -> {
-                BlockPos pos = pair.first();
-                if (pos.getY() >= 128) return;
-                boolean isSolid = pair.second() != AIR_BLOCK_STATE;
-                Octree.setBlock(ptr, pos.getX() & 15, pos.getY(), pos.getZ() & 15, isSolid);
-            });
+            // not inserting or deleting from the cache hashmap so we should be fine not being exclusive
```

Comment:
Except iirc C++ considers any data race undefined behavior and doing this across an ffi boundary won't make things better.

Reply:
It's undefined behavior if another thread is writing to the data at the same time we are reading it but that is not the case here.

Reply:
The pathfinder racing with this would be inconsequential however multiple calls to this to nearby blocks could cause bad data to be written so i will change it to use a write lock. The read executor is still single threaded so there currently isn't a difference anyways.